spec/edpm-openstack-service-architecture.md

```diff
@@ -0,0 +1,326 @@
+are defined in and therefore reconciled by the respective service operator
+(e.g. nova-operator, ovn-operator). These OpenStack service CRDs will be
+created automatically based on the existence of DataPlaneRole and Node CRs and
+the user specified annotations on these CRs. The service operator
+will ensure that both the podified control plane dependencies
+(e.g. message bus) are deployed and the EDPM node phase 0 is finished before
+generates the service configuration and creates AnsibleEE CRs to deploy
+OpenStack service on the EDPM node.
+
+The EDPM related OpenStack service CRDs will have `CustomServiceConfig` fields
+that will have the same syntax and semantic as the `CustomServiceConfig` fields
+in the existing podified Openstack service CRDs, while the implementation of
+such configuration will use Ansible to transfer the generated configuration to
+the service running on the EDPM node.
+
+**Benefits**:
+
+* The DataPlane CRDs and therefore the dataplane-operator implementation kept
+  OpenStack service agnostic. Adding support for deploying a new OpenStack
+  or partner service on the EDPM node can be done without changing the
+  DataPlane CRDs or the dataplane-operator implementation.
+
+* The service operators encapsulates all the service specific logic. Every
+  deployment, configuration or upgrade concerns for an OpenStack component can
+  be handled within a single, cohesive operator implementation.
+
+* The `CustomServiceConfig` field in the service CRDs simplifies and unifies
+  the OpenStack service configuration for the human deployer regardless where
+  the service is being deployed (i.e. podified or EDPM).
+
+**Drawbacks**:
+
+* The mechanism to watch DataPlane CRs from service operators and auto create
+  service CRs adds implementation complexity and introduces possible
+  performance impacts that could be avoided if dataplane-operator creates the
+  service CRs.
+
+* The Ready condition of the DataPlaneRole and the DataPlaneNode can only
+  represent the phase -1 and phase 0 readiness of the EDPM node. The phase 1
+  readiness can only be collected by checking the Ready condition of every
+  service CRs related to this EDPM node.
+
+
+### Example
+
+NovaCell/nova-cell1
+```yaml
+apiVersion: nova.openstack.org/v1beta1
+kind: NovaCell
+metadata:
+    name: nova-cell1
+...
+```
+No changes in NovaCell.
+
+OpenstackDataPlaneRole/edpm-compute
+```yaml
+apiVersion: dataplane.openstack.org/v1beta1
+kind: OpenStackDataPlaneRole
+metadata:
+    name: edpm-compute
+    annotations:
+        nova.openstack.org/cell: nova-cell1
+spec:
+    dataPlane: openstack-edpm
+    nodeTemplate:
+       ansiblePort: 22
+...
+```
+The `novaTemplate` is removed from the `nodeTemplate` and a new user specified
+annotation `nova.openstack.org/cell` is added so the human deployer can
+define that this Role describe a set of EDPM nodes that are compute nodes
+connected to cell1.
```

Comment:
ack so ``nova.openstack.org/cell: nova-cell1`` is the minium
information we would require to be able to generate a default novaExternalCompute CR
since it contain a reference to the nova cell by name which also uniquely identifies the Nova instance as the cell name contains the nova instance name as a prefix.

this would result in the nova/libvirt service being deploy with our default rendered config and attaching the node to cell1.